presto-main/src/main/java/com/facebook/presto/execution/SqlTaskManager.java

```diff
@@ -568,4 +580,42 @@ public QueryContext getQueryContext(QueryId queryId)
     {
         return queryContexts.getUnchecked(queryId);
     }
+
+    @Override
+    public BooleanIntImmutablePair shouldBackPressure(TaskUpdateRequest taskUpdateRequest)
+    {
+        if (!isTaskCreationBackpressureEnabled) {
+            return new BooleanIntImmutablePair(false, taskCreationBackpressureRetryAfter);
+        }
+
+        if (!taskUpdateRequest.getFirstTaskUpdate().isPresent() || !taskUpdateRequest.getFirstTaskUpdate().get()) {
+            return new BooleanIntImmutablePair(false, taskCreationBackpressureRetryAfter);
+        }
+
+        if (tasks.size() >= taskCreationTaskCountBackpressureThreshold) {
+            taskCreationBackpressureByTaskCount.update(1);
+            return new BooleanIntImmutablePair(true, taskCreationBackpressureRetryAfter);
+        }
+
+        if (Runtime.getRuntime().freeMemory() < taskCreationFreeMemoryLowBackpressureThreshold.toBytes()) {
```

Comment:
It seems this would be highly variable, since it's affected by when was the last GC.  And tuning this would be difficult, since it would require knowledge on what is a safe value that is not likely to be affected by GC fluctuations.